internal/providers/vault/provider.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/internal/tcache"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+var clients = tcache.NewCache[*vaultclient.Client]()
+
+func Register() {
+	combined.RegisterSecretsProvider(appRoleProvider)
+	combined.RegisterSecretsProvider(certificateProvider)
+}
+
+func login(ctx context.Context, caCfg *vault.VaultProvider, audience string) (*vaultclient.Client, error) {
+	key := caCfg.GetAddress() + caCfg.GetNamespace() + caCfg.GetAuthMount()
+	return clients.Compute(key, func() (*vaultclient.Client, error) {
```

Comment:
The usage of mutex would guarantee that the very first login would block all other for the same vault.
Would this be a case here?

Reply:
Yes, apparently the first thing that Compute does is [lock its own mutex](https://github.com/namespacelabs/foundation/blob/ba4497c865c56bc1c98bfb9768127cfed670b032/internal/tcache/tcache.go#L38-L50).